benchmark/src/test/java/org/reactivestreams/utils/benchmark/Benchmark.java

```diff
@@ -0,0 +1,182 @@
+            .filter(i -> i % 2 == 1)
+            .map(Function.identity())
+            .map(Function.identity())
+            .forEach(i -> count.incrementAndGet()).build(engine));
+        assert count.get() == iterations / 2;
+      });
+      System.out.println(complexProcessorLongStreamTime);
+
+      System.out.print(name + " one at a time materialization... ");
+      oneAtATimeMaterializationTime = time(() -> {
+        int iterations = 100000;
+        AtomicInteger count = new AtomicInteger();
+        Runnable[] runnable = new Runnable[1];
+        CompletableFuture<Void> done = new CompletableFuture<>();
+        runnable[0] = () -> {
+          if (count.getAndIncrement() < iterations) {
+            ReactiveStreams.of(1, 2, 3).forEach(e -> {}).build(engine)
+                .thenRunAsync(runnable[0]);
+          } else {
+            done.complete(null);
+          }
+        };
+        runnable[0].run();
+
+        await(done);
+      });
+      System.out.println(oneAtATimeMaterializationTime);
+
+      System.out.print(name + " concurrent materialization... ");
+      concurrentMaterializationTime = time(() -> {
+        int iterations = 100000;
+        int concurrency = 8;
+        Runnable[] runnable = new Runnable[concurrency];
+        CompletableFuture<Void>[] dones = new CompletableFuture[concurrency];
+        for (int i = 0; i < concurrency; i++) {
+          // effectively final...
+          int j = i;
+          AtomicInteger count = new AtomicInteger();
+          dones[j] = new CompletableFuture<>();
+          runnable[j] = () -> {
+            if (count.getAndIncrement() < iterations) {
+              ReactiveStreams.of(1, 2, 3).forEach(e -> {}).build(engine)
+                  .thenRunAsync(runnable[j]);
+            } else {
+              dones[j].complete(null);
+            }
+          };
+          runnable[j].run();
+        }
+        for (CompletableFuture<Void> done: dones) {
+          await(done);
+        }
+      });
+      System.out.println(concurrentMaterializationTime);
+
+      System.out.print(name + " sub stream time.... ");
+      subStreamTime = time(() -> {
+        int iterations = 100000;
+        AtomicInteger count = new AtomicInteger();
+        await(ReactiveStreams.fromIterable(() -> IntStream.range(1, iterations).iterator())
+            .flatMap(i -> ReactiveStreams.of(1, 2, 3, 4))
+            .forEach(i -> count.incrementAndGet()).build(engine));
+        assert count.get() == iterations * 4;
+      });
+      System.out.println(subStreamTime);
+    }
+  }
+
+  private static <T> T await(CompletionStage<T> future) {
+    try {
+      return future.toCompletableFuture().get(1, TimeUnit.MINUTES);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static long time(Runnable operation) {
+    System.gc();
+    long start = System.currentTimeMillis();
+    operation.run();
+    return System.currentTimeMillis() - start;
+  }
+}
```

Comment:
I'd suggest using jmh

Reply:
I don't intend on keeping this benchmarks here on the long term - I've never used jmh, I just wanted to do a quick comparison to make sure the performance of what I'd implemented wasn't terrible.

Reply:
When showing this to people, please do be sure to change to JMH before showing off as definitely it makes it look not serious at all (esp. as a proposal for JDK), and we need credibility here for this to take off.

JMH is very simple to use, please reach out if you have questions.

Reply:
So what's the best way to run a benchmark with three different backing engines? I can't see anything in the docs that allows for parameterization. Also, one of the main reasons why I did it myself is so that I could generate the nice looking table above, without that it's much harder to visualize the comparison. Does jmh support anything like that?

Reply:
Nice table of iffy measurements is the wrong side of things to optimize for...
Yes it has parameterization, `@Param`; https://github.com/akka/akka/blob/master/akka-bench-jmh/src/main/scala/akka/actor/ActorBenchmark.scala#L32
Yes, one can compare things, there are tools. https://github.com/akarnokd/jmh-compare-gui

Reply:
http://hg.openjdk.java.net/code-tools/jmh/file/25d8b2695bac/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_27_Params.java